pkg/controller/worker/machines.go

```diff
@@ -217,3 +264,340 @@ func (w *workerDelegate) generateMachineConfig(ctx context.Context) error {
+	err := w.ensureMigrationFirewall(ctx, metalControlPlane, infrastructureConfig, cluster, mclient, privateNetworkID)
+	if err != nil {
+		return err
+	}
+
+	return w.toggleAnnotation(ctx, cluster)
+}
+
+func (w *workerDelegate) ensureMigrationFirewall(ctx context.Context, metalControlPlane *apismetal.MetalControlPlane, infrastructureConfig *apismetal.InfrastructureConfig, cluster *extensionscontroller.Cluster, mclient metalgo.Client, privateNetworkID string) error {
+	var (
+		clusterID = string(cluster.Shoot.GetUID())
+		projectID = infrastructureConfig.ProjectID
+		namespace = cluster.ObjectMeta.Name
+	)
+
+	resp, err := mclient.Firewall().FindFirewalls(firewall.NewFindFirewallsParams().WithBody(&models.V1FirewallFindRequest{
+		AllocationProject: projectID,
+		Tags:              []string{clusterTag(clusterID)},
+	}).WithContext(ctx), nil)
+	if err != nil {
+		return fmt.Errorf("error finding firewall: %w", err)
+	}
+
+	var toMigrate []*models.V1FirewallResponse
+
+	for _, fw := range resp.Payload {
+		tm := tag.NewTagMap(fw.Tags)
+
+		if value, _ := tm.Value(fcmv2.FirewallControllerManagedByAnnotation); value == v2.FirewallControllerManager {
+			// firewall is already owned by the firewall-cotroller-manager, does not need migration
+			continue
+		}
+
+		toMigrate = append(toMigrate, fw)
+	}
+
+	if len(toMigrate) == 0 {
+		w.logger.Info("no firewalls to be migrated to firewall-controller-manager")
+		return nil
+	}
+
+	rateLimit := func(limits []apismetal.RateLimit) []fcmv2.RateLimit {
+		var result []fcmv2.RateLimit
+		for _, l := range limits {
+			result = append(result, fcmv2.RateLimit{
+				NetworkID: l.NetworkID,
+				Rate:      l.RateLimit,
+			})
+		}
+		return result
+	}
+
+	egressRules := func(egress []apismetal.EgressRule) []fcmv2.EgressRuleSNAT {
+		var result []fcmv2.EgressRuleSNAT
+		for _, rule := range egress {
+			rule := rule
+			result = append(result, fcmv2.EgressRuleSNAT{
+				NetworkID: rule.NetworkID,
+				IPs:       rule.IPs,
+			})
+		}
+		return result
+	}
+
+	fwcv, err := validation.ValidateFirewallControllerVersion(metalControlPlane.FirewallControllerVersions, infrastructureConfig.Firewall.ControllerVersion)
+	if err != nil {
+		return err
+	}
+
+	internalPrefixes := []string{}
+	if w.controllerConfig.AccountingExporter.Enabled && w.controllerConfig.AccountingExporter.NetworkTraffic.Enabled {
+		internalPrefixes = w.controllerConfig.AccountingExporter.NetworkTraffic.InternalNetworks
+	}
+
+	for _, fw := range toMigrate {
+		fw := fw
+
+		f := &fcmv2.Firewall{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      *fw.Allocation.Name,
+				Namespace: namespace,
+			},
+		}
+
+		_, err = controllerutil.CreateOrUpdate(ctx, w.client, f, func() error {
+			f.Labels = map[string]string{
+				tag.ClusterID: clusterID,
+			}
+			if v, err := semver.NewVersion(fwcv.Version); err == nil && v.LessThan(semver.MustParse("v2.0.0")) {
+				f.Annotations = map[string]string{
```

Comment:
Here we discard all annotations. Should we add `fcmv2.FirewallNoControllerConnectionAnnotation` instead? (check for nil)